### PR TITLE
feat(agents-api): add prefix to evaluate python expressions

### PR DESCRIPTION
**Pull request by Ahmad-mtos**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced a `$` prefix for evaluating Python expressions.

- Replaced `simple_eval_dict` with `base_evaluate` across multiple task steps.

- Added comprehensive tests for `base_evaluate` functionality.

- Removed deprecated `simple_eval_dict` utility and its tests.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>base_evaluate.py</strong><dd><code>Add `$` prefix support and enhance evaluation logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/activities/task_steps/base_evaluate.py

<li>Added support for <code>$</code> prefix in expression evaluation.<br> <li> Improved error handling for invalid expressions.<br> <li> Refactored evaluation logic to handle prefixed and non-prefixed <br>strings.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-05b346521c23f568c9804ac59931eb4fd5a61ec88110a783debb1ed329b47470">+5/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>evaluate_step.py</strong><dd><code>Replace `simple_eval_dict` with `base_evaluate`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/activities/task_steps/evaluate_step.py

<li>Replaced <code>simple_eval_dict</code> with <code>base_evaluate</code>.<br> <li> Updated logic to use the new evaluation method.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-ec34151f7ebbc4a46f47948fd5390ef4cac816e33011f86b37707b4e5e9f792c">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>log_step.py</strong><dd><code>Use `base_evaluate` for log evaluation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/activities/task_steps/log_step.py

<li>Replaced <code>render_template</code> with <code>base_evaluate</code> for log evaluation.<br> <li> Simplified evaluation logic for logging steps.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-b7ec416078e35c7f5e1e8fff666f9d8998a3da69b9fa5a1e10d21d19627e9cce">+2/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>prompt_step.py</strong><dd><code>Refactor prompt evaluation with `base_evaluate`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/activities/task_steps/prompt_step.py

<li>Removed <code>EVAL_PROMPT_PREFIX</code> and associated logic.<br> <li> Replaced <code>render_template</code> with <code>base_evaluate</code> for prompt evaluation.<br> <li> Simplified prompt evaluation for both strings and lists.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-568607bf9eff0a7b0bac363c2f9c64de9aec36b6e67497724377962151efb6f1">+13/-20</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>set_value_step.py</strong><dd><code>Update set value logic with `base_evaluate`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/activities/task_steps/set_value_step.py

<li>Replaced <code>simple_eval_dict</code> with <code>base_evaluate</code>.<br> <li> Updated logic for setting values using the new evaluation method.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-246c49c48648fa702bb110b8cc5ed51babcf94bc4cba9bb6989081611544efd2">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Remove deprecated `simple_eval_dict` utility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/activities/utils.py

<li>Removed deprecated <code>simple_eval_dict</code> utility.<br> <li> Cleaned up unused code related to expression evaluation.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-f0e36b882f3054b306bac13f5287f01af7b9d9c1dfb1a4c138966c8e333bc5fd">+0/-20</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_base_evaluate.py</strong><dd><code>Add tests for `base_evaluate` functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/tests/test_base_evaluate.py

<li>Added tests for <code>base_evaluate</code> covering strings, lists, and <br>dictionaries.<br> <li> Included edge cases like undefined values and empty expressions.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-6903501fdd4934d40307d1a19f64e8894d7b68754833ef9c6543c67cb0b66430">+42/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_simple_eval_dict.py</strong><dd><code>Remove tests for `simple_eval_dict`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/tests/test_simple_eval_dict.py

- Removed tests for the deprecated `simple_eval_dict` utility.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1107/files#diff-d0511ecc5dc47180b2055ae1d9aee2bc04bc5524fe884c9d4cb2b60488b5efd0">+0/-29</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

----

> [!IMPORTANT]
> Introduces `$` prefix for Python expression evaluation, replaces `simple_eval_dict` with `base_evaluate`, and updates workflows and tests accordingly.
> 
>   - **Behavior**:
>     - Introduced `$` prefix for evaluating Python expressions in `base_evaluate.py`.
>     - Replaced `simple_eval_dict` with `base_evaluate` in `evaluate_step.py`, `log_step.py`, and `prompt_step.py`.
>     - Removed `simple_eval_dict` utility from `utils.py`.
>   - **Workflow**:
>     - Updated `TaskExecutionWorkflow` to use `base_evaluate_activity` for expression evaluation.
>     - Refactored step handling logic to accommodate new evaluation method.
>   - **Tests**:
>     - Added tests for `base_evaluate` in `test_base_evaluate.py`.
>     - Removed tests for `simple_eval_dict` in `test_simple_eval_dict.py`.
>     - Updated workflow tests in `test_task_execution_workflow.py` to cover new evaluation logic.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=julep-ai%2Fjulep&utm_source=github&utm_medium=referral)<sup> for 2891535fcd97fdce14094ec06e2675faf06f4a0a. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: agents-api/agents_api/activities/task_steps/base_evaluate.py [hunk 1]
```diff
@@ -38,10 +38,14 @@ def __init__(self, error, expression, values):
 def _recursive_evaluate(expr, evaluator: SimpleEval):
     if isinstance(expr, str):
         try:
+            if isinstance(expr, str) and expr.startswith("$"):
```
Comment (creatorrr) [inline agents-api/agents_api/activities/task_steps/base_evaluate.py:41]: this should check for `"$ "` (notice space at the end). Because `$1million` could be a valid string and means something else entirely

**Comment by creatorrr** [comment]
shouldn't we simplify the steps and remove redundant ones? Meaning, we should run `base_evaluate` directly inside the workflow at the very beginning to get the evaluated definition and then proceed. This way we can remove most of the steps that just do an evaluate and nothing else. For example, [for_each_step.py](https://github.com/julep-ai/julep/blob/dev/agents-api/agents_api/activities/task_steps/for_each_step.py) doesn't do anything else.

File: agents-api/agents_api/activities/task_steps/base_evaluate.py [hunk 2]
```diff
@@ -57,9 +62,14 @@ def _recursive_evaluate(expr, evaluator: SimpleEval):
 @beartype
 async def base_evaluate(
     exprs: Any,
-    values: dict[str, Any] = {},
+    context: StepContext | None = None,
+    values: dict[str, Any] | None = None,
     extra_lambda_strs: dict[str, str] | None = None,
 ) -> Any | list[Any] | dict[str, Any]:
+    if context:
+        values = values or {} | (context and await context.prepare_for_step())
```
[the remark this answers is not in the document]
Comment (whiterabbit1983) [reply agents-api/agents_api/activities/task_steps/base_evaluate.py:70]: Also, what should be `values` equal to if `context` is `None` ?
